src/OrleansServiceBus/Providers/Streams/EventHub/EventHubCacheEvictionStrategy.cs

```diff
@@ -0,0 +1,142 @@
+                {
+                    break;
+                }
+                lastMessagePurged = oldestMessageInCache;
+                itemsPurged++;
+                this.PurgeObservable.RemoveOldestMessage();
+            }
+            var itemCountAfterPurge = itemCountBeforePurge - itemsPurged;
+
+            //purge finished, time to conduct follow up actions 
+            OnPurged?.Invoke(lastMessagePurged, this.PurgeObservable.Newest);
+            UpdatePurgedBuffers(lastMessagePurged, this.PurgeObservable.Oldest, itemsPurged > 0);
+            ReportPurge(itemCountBeforePurge, itemCountAfterPurge);
+        }
+
+        private void UpdatePurgedBuffers(CachedEventHubMessage? lastMessagePurged, CachedEventHubMessage? oldestMessageInCache, bool itemsGotPurged)
+        {
+            //if nothing purged, then no buffer was purged
+            //if items got purged, then potencially some buffer was purged
+            if (itemsGotPurged)
+            {
+                var IdOfLastPurgedBuffer = lastMessagePurged.Value.Segment.Array;
+                // IdOfLastBufferInCache will be null if cache is empty after purge
+                var IdOfLastBufferInCache = oldestMessageInCache.HasValue ? oldestMessageInCache.Value.Segment.Array : null;
+                //all buffer older than LastPurgedBuffer should be purged 
+                while (this.inUseBuffers.Peek().Id != IdOfLastPurgedBuffer)
+                {
+                    this.purgedBuffers.Enqueue(this.inUseBuffers.Dequeue());
+                }
+                // if last purged message does not share buffer with remaining messages in cache, or cache is empty after purge, 
+                //then last purged buffer should be in purgedBuffers too
+                if (IdOfLastBufferInCache == null || IdOfLastPurgedBuffer != IdOfLastBufferInCache)
+                {
+                   this.purgedBuffers.Enqueue(this.inUseBuffers.Dequeue());
+                }
+            }
+        }
+
+        // Given a purge cached message, indicates whether it should be purged from the cache
+        private bool ShouldPurge(ref CachedEventHubMessage cachedMessage, ref CachedEventHubMessage newestCachedMessage, DateTime nowUtc)
+        {
+            TimeSpan timeInCache = nowUtc - cachedMessage.DequeueTimeUtc;
+            // age of message relative to the most recent event in the cache.
+            TimeSpan relativeAge = newestCachedMessage.EnqueueTimeUtc - cachedMessage.EnqueueTimeUtc;
+
+            return timePurge.ShouldPurgFromTime(timeInCache, relativeAge);
+        }
+
+        private void OnFreeBlockRequest(IDisposable block)
+        {
+            var purgeCandidate = block as FixedSizeBuffer;
+            //free all blocks before purgeCandidate and purgeCandidate
+            if (this.purgedBuffers.Contains(purgeCandidate))
+            {
+                while (this.purgedBuffers.Peek() != purgeCandidate)
+                    this.purgedBuffers.Dequeue().Dispose();
```

Comment:
shouldn't we dequeue and dispose every buffer and break after we dispose the purgeCandidate?  This logic will leave the purgeCandidate in the queue, which will double dispose it.